src/platform/ASR/ASRFactoryDataParser.c

```diff
@@ -0,0 +1,308 @@
+#define asr_flash_read duet_flash_read
+#define asr_factory_decrypt_dac_prvkey duet_factory_decrypt_dac_prvkey
+#else
+#include "lega_cm4.h"
+#include "lega_flash.h"
+#define asr_flash_read lega_flash_read
+#endif
+
+#if CONFIG_ENABLE_ASR_FACTORY_DATA_PROVIDER
+
+/* Logic partition on flash devices for matter */
+const matter_partition_t asr_matter_partitions_table[] =
+{
+    [ASR_VERSION_PARTITION] =
+    {
+        .partition_name = "version",
+    },
+    [ASR_CONFIG_PARTITION] =
+    {
+        .partition_name = "config",
+    },
+    [ASR_ITERATION_COUNT_PARTITION] =
+    {
+        .partition_name = "iteration-count",
+    },
+    [ASR_SALT_PARTITION] =
+    {
+        .partition_name = "salt",
+    },
+    [ASR_VERIFIER_PARTITION] =
+    {
+        .partition_name = "verifier",
+    },
+    [ASR_DISCRIMINATOR_PARTITION] =
+    {
+        .partition_name = "discriminator",
+    },
+    [ASR_DAC_CERT_PARTITION] =
+    {
+        .partition_name = "dac-cert",
+    },
+    [ASR_DAC_KEY_PARTITION] =
+    {
+        .partition_name = "dac-pri-key",
+    },
+    [ASR_DAC_PUB_KEY_PARTITION] =
+    {
+        .partition_name = "dac-pub-key",
+    },
+    [ASR_PAI_CERT_PARTITION] =
+    {
+        .partition_name = "pai-cert",
+    },
+    [ASR_CERT_DCLRN_PARTITION] =
+    {
+        .partition_name = "cert-dclrn",
+    },
+    [ASR_CHIP_ID_PARTITION] =
+    {
+        .partition_name = "chip-id",
+    },
+#if CONFIG_ENABLE_ASR_FACTORY_DEVICE_INFO_PROVIDER
+    [ASR_VENDOR_NAME_PARTITION] =
+    {
+        .partition_name = "vendor-name",
+    },
+    [ASR_VENDOR_ID_PARTITION] =
+    {
+        .partition_name = "vendor-id",
+    },
+    [ASR_PRODUCT_NAME_PARTITION] =
+    {
+        .partition_name = "product-name",
+    },
+    [ASR_PRODUCT_ID_PARTITION] =
+    {
+        .partition_name = "product-id",
+    },
+    [ASR_ROTATING_UNIQUE_ID_PARTITION] =
+    {
+        .partition_name = "rd-id-uid",
+    },
+    [ASR_MANUFACTURY_DATE_PARTITION] =
+    {
+        .partition_name = "mfg-date",
+    },
+    [ASR_SERIAL_NUMBER_PARTITION] =
+    {
+        .partition_name = "serial-num",
+    },
+    [ASR_HARDWARE_VERSION_PARTITION] =
+    {
+        .partition_name = "hardware-ver",
+    },
+    [ASR_HARDWARE_VERSION_STR_PARTITION] =
+    {
+        .partition_name = "hw-ver-str",
+    },
+    [ASR_PRODUCT_URL_PARTITION] =
+    {
+        .partition_name = "product-url",
+    },
+    [ASR_PRODUCT_LABEL_PARTITION] =
+    {
+        .partition_name = "product-label",
+    },
+    [ASR_PART_NUMBER_PARTITION] =
+    {
+        .partition_name = "part-number",
+    },
+#endif
+    [ASR_MATTER_PARTITION_MAX] =
+    {
+        .partition_name = NULL, //for end don't change,
+    }
+};
+
+static asr_tlv_context matter_tlv_ctx;
+
+uint32_t asr_matter_find_by_name(const char * name, uint8_t * buf, uint32_t buf_len, uint32_t * out_len)
+{
+    tlv_header_t tlv;
+
+    if (asr_tlv_find_by_name(&matter_tlv_ctx, &tlv, MIXCLASS_MATTERCONFIG_TAG,
```

Comment:
you could reduce overall intend by:

```
if (asr_tlv_find_by_name(/*...*/) == 0) {
    return -1;
}

uint32_t value_len = tlv_htons(tlv->data_len);
//...
```

please add constants for -1, -2 and maybe even 0 to mean ok. 

Overall API comment: it seems quite odd that the tlv find_by_name returns a bool and the asr_matter one returns an error code (basically switching the meaning of 0). I could not find documentation so guessing this is intentional, but seems very error prone.

Reply:
all return changs to use constants